pkg/images/set_test.go

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2020-Present Pivotal Software, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package images_test
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/pivotal/image-relocation/pkg/image"
+	"github.com/pivotal/image-relocation/pkg/images"
+)
+
+var _ = Describe("FlattenRepoPath", func() {
+	Describe("New", func() {
+		var (
+			ii  []string
+			s   images.Set
+			err error
+		)
+
+		JustBeforeEach(func() {
+			s, err = images.New(ii...)
+		})
+
+		Context("when the input slice is empty", func() {
+			BeforeEach(func() {
+				ii = []string{}
+			})
+
+			It("should construct an empty set", func() {
+				Expect(err).NotTo(HaveOccurred())
+				Expect(s).To(Equal(images.Empty))
+			})
+		})
+
+		Context("when the input slice is non-empty", func() {
+			BeforeEach(func() {
+				ii = []string{"a", "b"}
+			})
+
+			It("should construct the correct set", func() {
+				Expect(err).NotTo(HaveOccurred())
+
+				v, err := images.New("b", "a")
+				Expect(err).NotTo(HaveOccurred())
+
+				Expect(s).To(Equal(v))
+			})
+		})
+
+		Context("when the input slice has an invalid image reference", func() {
+			BeforeEach(func() {
+				ii = []string{"::"}
+			})
+
+			It("should return a suitable error", func() {
+				Expect(err).To(MatchError(`invalid image reference: "::"`))
+			})
+		})
+	})
+
+	Describe("Union", func() {
```

Comment:
property-based testing seems to be an ideal fit for this kind of invariant verification (instead of hardcoding a few simple example cases like below)
you can use the built-in [`testing/quick`](https://golang.org/pkg/testing/quick/) package for that

Reply:
Thanks for this idea. I had a think about it and I'm not sure it would be worth the extra complexity in the tests. In particular, we're not really talking about testing invariants so much as testing postconditions. So the inputs need to be particular so that their union is as expected. It's not clear how generating randomised sets would help. Of course, we could generate randomised image references but we can use the examples here without loss of generality.